models/hh_cond_beta_gap_traub.cpp

```diff
@@ -0,0 +1,787 @@
+
+void
+nest::hh_cond_beta_gap_traub::init_state_( const Node& proto )
+{
+  const hh_cond_beta_gap_traub& pr =
+    downcast< hh_cond_beta_gap_traub >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::hh_cond_beta_gap_traub::init_buffers_()
+{
+  B_.spike_exc_.clear(); // includes resize
+  B_.spike_inh_.clear(); // includes resize
+  B_.currents_.clear();  // includes resize
+
+  // allocate strucure for gap events here
+  // function is called from Scheduler::prepare_nodes() before the
+  // first call to update
+  // so we already know which interpolation scheme to use according
+  // to the properties of this neurons
+  // determine size of structure depending on interpolation scheme
+  // and unsigned int Scheduler::min_delay() (number of simulation time steps
+  // per min_delay step)
+
+  // resize interpolation_coefficients depending on interpolation order
+  const size_t buffer_size = kernel().connection_manager.get_min_delay()
+    * ( kernel().simulation_manager.get_wfr_interpolation_order() + 1 );
+
+  B_.interpolation_coefficients.resize( buffer_size, 0.0 );
+
+  B_.last_y_values.resize( kernel().connection_manager.get_min_delay(), 0.0 );
+
+  B_.sumj_g_ij_ = 0.0;
+
+  Archiving_Node::clear_history();
+
+  B_.logger_.reset();
+
+  B_.step_ = Time::get_resolution().get_ms();
+  B_.IntegrationStep_ = B_.step_;
+
+  if ( B_.s_ == 0 )
+  {
+    B_.s_ =
+      gsl_odeiv_step_alloc( gsl_odeiv_step_rkf45, State_::STATE_VEC_SIZE );
+  }
+  else
+  {
+    gsl_odeiv_step_reset( B_.s_ );
+  }
+
+  if ( B_.c_ == 0 )
+  {
+    B_.c_ = gsl_odeiv_control_y_new( 1e-3, 0.0 );
+  }
+  else
+  {
+    gsl_odeiv_control_init( B_.c_, 1e-3, 0.0, 1.0, 0.0 );
+  }
+
+  if ( B_.e_ == 0 )
+  {
+    B_.e_ = gsl_odeiv_evolve_alloc( State_::STATE_VEC_SIZE );
+  }
+  else
+  {
+    gsl_odeiv_evolve_reset( B_.e_ );
+  }
+
+  B_.sys_.function = hh_cond_beta_gap_traub_dynamics;
+  B_.sys_.jacobian = 0;
+  B_.sys_.dimension = State_::STATE_VEC_SIZE;
+  B_.sys_.params = reinterpret_cast< void* >( this );
+
+  B_.I_stim_ = 0.0;
+}
+
+double
+nest::hh_cond_beta_gap_traub::get_normalisation_factor( double tau_rise,
+  double tau_decay )
+{
+  // Factor used to normalise the synaptic conductance such that
+  // incoming spike causes a peak conductance of 1 nS.
+  // The denominator (denom1) that appears in the expression of the peak time
+  // is computed here to check that it is != 0
+  // another denominator denom2 appears in the expression of the
+  // normalization factor g0
+  // Both denom1 and denom2 are null if tau_decay = tau_rise, but they
+  // can also be null if tau_decay and tau_rise are not equal but very
+  // close to each other, due to the numerical precision limits.
+  // In such case the beta function reduces to the alpha function,
+  // and the normalization factor for the alpha function should be used.
+  double denom1 = tau_decay - tau_rise;
+  double denom2 = 0;
+  double normalisation_factor = 0;
+  if ( denom1 != 0 )
```

Comment:
I suggest replacing the condition with `|denom1| < std::numeric_limits< double >::epsilon()`

Also: why is denom2 calculated inside the first if block? Can we use the return-early pattern here?

Reply:
I agree with the suggestion to check ```std::abs(denom1) < std::numeric_limits< double >``` makes sense, also applies to the check for `denom2`.

Reply:
Hi, this section of the code is the same as aeif_cond_beta_multisynapse, so I suggest these changes should also apply there.